### DISCUSSION TITLE
Reading .dat files using external reader

### DISCUSSION THREAD

**Discussion by Ramogi254** [ansys/pyfluent#2047] [Q&A]
Hello,

I need some assistance regarding how to read a .dat Ansys Fluent simulation data file using an external reader - python. The steps below are the specifics of what I would like to achieve. 

Step 1: Getting the solution data in .dat format

Step 2: Read the solution data file using python

Step 3: Modify the solution using Singular Value Decomposition

Step 4: Reconstruct the solution using a few mode coefficients

Step 5: Write the reconstructed solution from python code into another .dat file

Step 6: Read reconstructed solution into CFD post.

Step 7: Visualize the data.

My main concern is how to read and write the ANSYS Fluent solution data whose format is unique (steps 2 and 5). I would appreciate any assistance with regards to this.

Thanks,
Emmanuel

**Comment by RobPasMue** [comment]
@ansys/pyfluent-contributors - can we give this user some assistance? He also contacted @ansys/pyansys-core via email. Thanks!

**Comment by seanpearsonuk** [comment]
@Ramogi254 .dat files that are needed can be generated in Fluent in the usual. `class DataFile` is documented in data_file.py, but not yet in the documentation website. data_file.py contains sufficient basic information to interact with the class to read the data. The website will contain more comprehensive infomation.

@ajain-work 
@RobPasMue

**Comment by Ramogi254** [comment]
@seanpearsonuk I have already generated the .dat file in Fluent. The steps that I really need assistance with are steps 2 & 5. Can you share the link to the documentation website you have mentioned? And also a link for data_file.py? Thanks.

**Comment by seanpearsonuk** [comment]
@Ramogi254 

https://github.com/ansys/pyfluent/blob/main/src/ansys/fluent/core/filereader/data_file.py
https://fluent.docs.pyansys.com/

Step 5 is not a feature

**Reply by Ramogi254** [reply]
@seanpearsonuk thank you for the swift replies. 

I run the code provided in data_file.py (shown below)
`from ansys.fluent.core import examples
from ansys.fluent.core.filereader.datareader import DataReader`

However I get the following error.
`ModuleNotFoundError: No module named 'ansys.fluent.core.filereader.datareader'`

Is there a way to install this particular module? Or is there another way to call the module?

PS: I launched Jupyter notebook using the Ansys Python Manager.

Thanks

**Comment by seanpearsonuk** [comment]
@Ramogi254 I have made some comments [in an issue here](https://github.com/ansys/pyfluent/issues/2046#issuecomment-1741261805)



Please refer to my latest suggestion there.

**Reply by Ramogi254** [reply]
@seanpearsonuk I'm not sure why but the same error seems to pop up. The changed module cannot be found.

**Reply by seanpearsonuk** [reply]
@Ramogi254 maybe send a screenshot

**Reply by Ramogi254** [reply]
![ModuleNotFoundError](https://github.com/ansys/pyfluent/assets/145783999/2d8032a3-eff4-45f9-8d7c-014d6eae187d)

**Reply by raph-luc** [reply]
@Ramogi254 @seanpearsonuk I believe that feature hasn't made it to release yet, latest versions include changes up to August 29, and the changes to `data_file` were committed September 4. Need to clone the PyFluent repo and install development version if you want to make use of `data_file`:  https://fluent.docs.pyansys.com/version/stable/getting_started/installation.html

**Reply by Ramogi254** [reply]
@raph-luc I have tried using the second batch of code on the link you sent above (I believe it must be the one for cloning the PyFluent repo). However, each line of code brings up a syntax error. Here is a screenshot of what I mean. 
![image](https://github.com/ansys/pyfluent/assets/145783999/2e9ecf32-1738-4ac9-9052-3dd256ecfe3b)

**Comment by raph-luc** [comment]
@Ramogi254 The commands in that installation guide are not Python commands, you need to open a system command shell or terminal to follow those steps, i.e. Windows CMD if you are using Windows.

And yes, you are looking at the correct instructions, the commands are:
```
git clone https://github.com/ansys/pyfluent.git
cd pyfluent
pip install pip -U
pip install -e .
python codegen/allapigen.py 
```

**Comment by Ramogi254** [comment]
@raph-luc I have successfully run the first 3 lines of code. I was running the second last line of code `pip install -e .` and came across this error in Command Prompt. How can I correct this?
![cmd error](https://github.com/ansys/pyfluent/assets/145783999/c6d43b6b-901f-4c79-91d2-bfde1dd7adc1)

**Reply by raph-luc** [reply]
I unfortunately can't help you with that one, I've never encountered that error. According to the error message, seems that you may need to manually install HDF5 in your system?

**Reply by Ramogi254** [reply]
@raph-luc I have installed it manually however the same error pops up. I was just wondering, in case you can help, where to move the directory files if at all that can help. 
![image](https://github.com/ansys/pyfluent/assets/145783999/8f463a5c-cdb2-4601-a95d-33e7817a6263)

NB: I had to install HDF5 files using the Windows subsystem for linux.

**Reply by raph-luc** [reply]
@Ramogi254 New PyFluent version updates have just been released that should achieve what you need as it includes the `DataFile` changes, you can install the released version using:
`pip install ansys-fluent-core==0.18.dev1`

**Reply by mkundu1** [reply]
@Ramogi254 You may see this issue if you're using Python version 3.12. We don't support 3.12 yet (open issue - #2079).
